ckanext/harvest/logic/action/update.py

```diff
@@ -605,106 +612,156 @@ def harvest_jobs_run(context, data_dict):
-        msg += toolkit._('Error Summary')
-        msg += '\n'
-
-    if (obj_error != ''):
-        msg += toolkit._('Document Error')
-        msg += '\n' + obj_error + '\n\n'
-
-    if (job_error != ''):
-        msg += toolkit._('Job Errors')
-        msg += '\n' + job_error + '\n\n'
-
-    if obj_error or job_error:
-        msg += '\n--\n'
-        msg += toolkit._('You are receiving this email because you are currently set-up as Administrator for {0}.'
-                         ' Please do not reply to this email as it was sent from a non-monitored address.')\
-            .format(config.get('ckan.site_title'))
-
-        recipients = []
-
-        # gather sysadmins
-        model = context['model']
-        sysadmins = model.Session.query(model.User).filter(
-            model.User.sysadmin == True  # noqa: E712
-        ).all()
-        for sysadmin in sysadmins:
-            recipients.append({
-                'name': sysadmin.name,
-                'email': sysadmin.email
-            })
-
-        # gather organization-admins
-        if source.get('organization'):
-            members = get_action('member_list')(context, {
-                'id': source['organization']['id'],
-                'object_type': 'user',
-                'capacity': 'admin'
-            })
-            for member in members:
-                member_details = get_action('user_show')(context, {'id': member[0]})
-                if member_details['email']:
-                    recipients.append({
-                        'name': member_details['name'],
-                        'email': member_details['email']
-                    })
-
-        for recipient in recipients:
-            email = {'recipient_name': recipient['name'],
-                     'recipient_email': recipient['email'],
-                     'subject': config.get('ckan.site_title') + ' - Harvesting Job - Error Notification',
-                     'body': msg}
+    # gather sysadmins
+    model = context['model']
+    sysadmins = model.Session.query(model.User).filter(
+        model.User.sysadmin == True  # noqa: E712
+    ).all()
+
+    for sysadmin in sysadmins:
+        recipients.append({
+            'name': sysadmin.name,
+            'email': sysadmin.email
+        })
+
+    # gather organization-admins
+    if source.get('organization'):
+        members = get_action('member_list')(context, {
+            'id': source['organization']['id'],
+            'object_type': 'user',
+            'capacity': 'admin'
+        })
+
+        for member in members:
+            member_details = get_action(
+                'user_show')(context, {'id': member[0]})
+
+            if member_details['email']:
+                recipients.append({
+                    'name': member_details['name'],
+                    'email': member_details['email']
+                })
+
+    # allow plugins to add custom recipients 
+    for p in PluginImplementations(IActions):
```

Comment:
The standard way of extending actions is to use [chained actions](https://docs.ckan.org/en/2.9/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IActions.get_actions). ckanext-harvest would implement the main `harvest_add_extra_notification_recipients` (better to prefix it with `harvest_` to avoid conflicts), which would replace the current `get_recipients`.
Then other extension would implement their own action as:

```python
@toolkit.chained_action
def harvest_add_extra_notification_recipients(up_func, context, data_dict):
    recipients = up_func(context, data_dict)
    # Custom logic
    recipients += new_recipients
    return recipients
```

Does this make sense?

Reply:
@amercader we are [using this action](https://github.com/GSA/ckanext-datagovcatalog/blob/master/ckanext/datagovcatalog/harvester/notifications.py#L9) from a custom extension (not a harvester). This extension analyzes `extras` from an `organization` to look for new recipients.
So, maybe this is not only for harvesters. Probably users will want to define extra recipients in configuration or in other extensions. I think _recipients_ is not just about harvesters

Anyway, [I try to create a new branch](https://github.com/avdata99/ckanext-harvest/commit/f59e373341cb31c2ba340cc1d340d64963a4670c) to transform this into a `chained_action`. Is this a good approach? I'm not sure.
If we use this: will we need to use in all harvester or we will be able to use from a non-harvester extension?

Reply:
@avdata99 I think actions should be explicit in what they do and be granular. So my suggested approach would be:

In **ckanext-harvest**:

In *action/get.py*:

```python
# This will automatically get registered as action, no need to explicitly register it in plugin.py
def harvest_get_notifications_recipients(context, data_dict):
   # See auth/get.py below
   check_access('harvest_get_notifications_recipients', contex, data_dict)

   # Include here the default logic for getting recipients (what you have in this PR minus the lines below to call other actions)
    return recipients

```

In *auth/get.py*:

```python

def harvest_get_notifications_recipients(context, data_dict):
   # Only sysadmins can access this
    return {'success': False}
```

In *action/update.py*:

```python

def send_summary_email(context, source_id, status):
    subject, body = prepare_summary_mail(context, source_id, status)
    # Using get_action here will ensure that if another extension implements this action it will be called (if loaded first in the ini file)
    recipients = toolkit.get_action('harvest_get_notifications_recipients')(context, {'source_id': source_id})
    send_mail(recipients, subject, body)

def send_error_email(context, source_id, status):
    subject, body = prepare_error_mail(context, source_id, status)
    recipients = toolkit.get_action('harvest_get_notifications_recipients')(context, {'source_id': source_id})
    send_mail(recipients, subject, body)

```

In **ckanext-datagovcatalog**:

In *actions.py* (or whatever file depending on how you structured your extension):

Option 1, if you want to extend the default behaviour from ckanext-harvest, use a chained action:

```python
@toolkit.chained_action
def harvest_get_notifications_recipients(up_func, context, data_dict):
     # This calls the ckanext-harvest parent function
     recipients = up_func(context, data_dict)

    # Do any extra stuff to modify the default recipients
    
     return recipients
```

Option 2, If you want another behaviour, just override the whole function:
```python
def harvest_get_notifications_recipients(context, data_dict):

    # Do your custom stuff
    
     return recipients
```

If you have a generic `get_recipients()` function which is used elsewhere on your extension for other notifcations just call it from the overridden action:

 ```python
def harvest_get_notifications_recipients(context, data_dict):

    recipients = get_recipients()
    
     return recipients
```

Does this makes sense? I think that makes the implementation simpler and cleaner

Reply:
Ok @amercader, I make these changes [here](https://github.com/ckan/ckanext-harvest/pull/407/commits/c84f63896c6e483e17df9fc10c6d22ff6ec94dd5).
I will update my extension to fit this